File: framework/languages/dotnet/e2e/ResilienceLifecycle/Client/Support/ClientOptions.cs
namespace ResilienceLifecycle.Client.Support;

internal sealed record ClientOptions(
    string ConsumerUrl,
    string TopologyUrl,
    string RedisEndpoint,
    string RedisKeyPrefix,
    string RedisContainer,
    string ProviderAUrl,
    string ProviderAEndpoint,
    string ProviderAEvidenceFile,
    string ProviderBUrl,
    int ProviderBProcessId,
    string ProviderBEndpoint,
    string ProviderBEvidenceFile,
    string ProviderBRemapUrl,
    string ProviderBRemapEndpoint,
    string ProviderBGreenUrl,
    string ProviderBGreenEndpoint,
    string ProviderProject,
    string LogDir,
    string Scenario)
{
    public static ClientOptions Parse(string[] args)
    {
        var values = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
        for (var i = 0; i < args.Length; i++)
        {
            var key = args[i];
            if (!key.StartsWith("--", StringComparison.Ordinal)) continue;

            if (i + 1 >= args.Length) throw new ArgumentException($"Missing value for {key}.");

            values[key[2..]] = args[++i];
        }

        return new ClientOptions(
            values.TryGetValue("consumer-url", out var consumerUrl) && !string.IsNullOrWhiteSpace(consumerUrl)
                ? consumerUrl
                : throw new ArgumentException("--consumer-url is required."),
            Require(values, "topology-url"),
            Require(values, "redis-endpoint"),
            Require(values, "redis-key-prefix"),
            Require(values, "redis-container"),
            values.TryGetValue("provider-a-url", out var providerAUrl) && !string.IsNullOrWhiteSpace(providerAUrl)
                ? providerAUrl
                : throw new ArgumentException("--provider-a-url is required."),
            Require(values, "provider-a-endpoint"),
            Require(values, "provider-a-evidence-file"),
            values.TryGetValue("provider-b-url", out var providerBUrl) && !string.IsNullOrWhiteSpace(providerBUrl)
                ? providerBUrl
                : throw new ArgumentException("--provider-b-url is required."),
            int.TryParse(Require(values, "provider-b-process-id"), out var providerBProcessId)
                ? providerBProcessId
                : throw new ArgumentException("--provider-b-process-id must be an integer."),
            Require(values, "provider-b-endpoint"),
            Require(values, "provider-b-evidence-file"),
            Require(values, "provider-b-remap-url"),
            Require(values, "provider-b-remap-endpoint"),
            Require(values, "provider-b-green-url"),
            Require(values, "provider-b-green-endpoint"),
            Require(values, "provider-project"),
            Require(values, "log-dir"),
            values.TryGetValue("scenario", out var scenario) && !string.IsNullOrWhiteSpace(scenario)
                ? scenario
                : "all");
    }

    private static string Require(IReadOnlyDictionary<string, string> values, string name)
    {
        return values.TryGetValue(name, out var value) && !string.IsNullOrWhiteSpace(value)
            ? value
            : throw new ArgumentException($"--{name} is required.");
    }
}

File: framework/languages/node/e2e/ResilienceLifecycle/Client/Support/client-options.ts
export interface ClientOptions {
  readonly peerLocationUrl: string;
  readonly providerAUrl: string;
  readonly providerBUrl: string;
  readonly consumerUrl: string;
  readonly consumerUrls: readonly string[];
  readonly soakDurationMs: number;
  readonly redisEndpoint: string;
  readonly redisKeyPrefix: string;
  readonly redisContainer: string;
  readonly providerAChannelEndpoint: string;
  readonly providerBChannelEndpoint: string;
  readonly providerBRemapUrl: string;
  readonly providerBRemapChannelEndpoint: string;
  readonly providerBGreenUrl: string;
  readonly providerBGreenChannelEndpoint: string;
  readonly providerMain: string;
  readonly logDir: string;
  readonly scenario: string;
}

export function parseClientOptions(args: readonly string[]): ClientOptions {
  const values = new Map<string, string>();
  for (let i = 0; i < args.length; i += 1) {
    const key = args[i];
    if (!key.startsWith('--')) {
      continue;
    }
    if (i + 1 >= args.length) {
      throw new Error(`Missing value for ${key}.`);
    }
    values.set(key.slice(2), args[++i]);
  }
  const soakDurationSeconds = Number(required(values, 'soak-duration-seconds'));
  if (!Number.isInteger(soakDurationSeconds) || soakDurationSeconds < 120) {
    throw new Error('--soak-duration-seconds must be an integer of at least 120.');
  }
  return {
    peerLocationUrl: required(values, 'peer-location-url'),
    providerAUrl: required(values, 'provider-a-url'),
    providerBUrl: required(values, 'provider-b-url'),
    consumerUrl: required(values, 'consumer-url'),
    consumerUrls: required(values, 'consumer-urls').split(',').filter((value) => value.length > 0),
    soakDurationMs: soakDurationSeconds * 1000,
    redisEndpoint: required(values, 'redis-endpoint'),
    redisKeyPrefix: required(values, 'redis-key-prefix'),
    redisContainer: required(values, 'redis-container'),
    providerAChannelEndpoint: required(values, 'provider-a-channel-endpoint'),
    providerBChannelEndpoint: required(values, 'provider-b-channel-endpoint'),
    providerBRemapUrl: required(values, 'provider-b-remap-url'),
    providerBRemapChannelEndpoint: required(values, 'provider-b-remap-channel-endpoint'),
    providerBGreenUrl: required(values, 'provider-b-green-url'),
    providerBGreenChannelEndpoint: required(values, 'provider-b-green-channel-endpoint'),
    providerMain: required(values, 'provider-main'),
    logDir: required(values, 'log-dir'),
    scenario: values.get('scenario') ?? 'all'
  };
}

function required(values: ReadonlyMap<string, string>, key: string): string {
  const value = values.get(key);
  if (value === undefined || value.length === 0) {
    throw new Error(`--${key} is required.`);
  }
  return value;
}
